Add Steal Time support in ARM

DakshinD opened the pull request:
Opening this PR as a draft as we continue working on this issue to add steal time support for ARM. Please see below for the changes we have implemented so far.

## Changes

- Added PVTime struct in `src/vmm/src/arch/aarch64/pvtime/mod.rs` that allocates and stores per-vCPU stolen_time regions
  - It offers a `new` constructor to create the device (allocating system memory and storing regions per vcpu), and `register_vcpu` which attempts to use `kvm_set_device_attr` to register the corresponding IPA for a vcpu
- Created a function `setup_pv_time` in `src/vmm/src/builder.rs` which is invoked in `build_microvm_for_boot`
  - Checks if PVTime is supported using `kvm_has_device_attr`
  - Creates a PVTime device
  - Iterates over vcpu fd's to register their IPAs for steal_time struct

- Started using debug/info logs while booting microvm for testing

### Current problems:

 
### Future Progress
- We hope to continue making progress on this PR by:
  - Getting feedback on our current design of the PVTime device and `setup_pv_time`
  - Support snapshotting
  - Any other things we need to address

## Reason

Attempting to resolve issue #4866 

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] I have read and understand [CONTRIBUTING.md][3].
- [x] I have run `tools/devtool checkstyle` to verify that the PR passes the
  automated style checks.
- [ ] I have described what is done in these changes, why they are needed, and
  how they are solving the problem in a clear and encompassing way.
- [ ] I have updated any relevant documentation (both in code and in the docs)
  in the PR.
- [ ] I have mentioned all user-facing changes in `CHANGELOG.md`.
- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] When making API changes, I have followed the
  [Runbook for Firecracker API changes][2].
- [ ] I have tested all new and changed functionalities in unit tests and/or
  integration tests.
- [ ] I have linked an issue to every new `TODO`.

______________________________________________________________________

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

Manciukic commented on src/vmm/src/arch/aarch64/pvtime/mod.rs:
```diff
@@ -0,0 +1,117 @@
+    ) -> Result<Self, PVTimeError> {
+        info!("Creating PVTime for {vcpu_count} vCPUs...");
+        // This returns the IPA(?) of the start of our shared memory region for all vCPUs.
+        // Q: Confirm that allocate_system_memory returns an IPA?
```
yes, the return of this call is an IPA, iow a guest physical address. You may use the type `GuestAddress` (a wrapper around `u64`) to make this clear.

Manciukic commented on src/vmm/src/arch/aarch64/pvtime/mod.rs:
```diff
@@ -0,0 +1,117 @@
+        let vcpu_device_attr = kvm_bindings::kvm_device_attr {
+            group: KVM_ARM_VCPU_PVTIME_CTRL,
+            attr: KVM_ARM_VCPU_PVTIME_IPA as u64,
+            addr: ipa_ptr as u64, // userspace address of attr data
```
The "IPA" ("Intermediate Physical Address" or simply Guest Physical Address, gpa) is already an address in the guest physical memory address space, so it should be passed directly.

This may be confusing as, when dealing with virtualization, there's not just one view of the memory (address space), but (at least) 4: virtual and physical addresses from guest and host. In this case 2 are important:
 - GPA: guest physical address, ie the physical view of the system from the point of view of the guest (VM). The `ipa` is an address in this space.
 - HVA: host virtual address, ie the address space of the hypervisor process (Firecracker). The `ipa_ptr` is an address in this space as it's a pointer to a memory location in FC virtual memory.

For a more detailed explanation:
 - https://www.alibabacloud.com/blog/599058
 - https://kernel.org/doc/Documentation/virtual/kvm/mmu.txt

Manciukic reviewed (changes requested):
Let's try making the proposed change to pass the right address and see if it works. 
I haven't left any comment regarding style or design to focus on functionality first. Once we get something working we can see how to make it "look better".

Manciukic commented on src/vmm/src/arch/aarch64/pvtime/mod.rs:
```diff
@@ -0,0 +1,117 @@
+use crate::device_manager::resources::ResourceAllocator;
+
+/// Size of the stolen_time struct in bytes, see 3.2.2 in DEN0057A
+pub const STEALTIME_STRUCT_MEM_SIZE: u64 = 16;
```
although it requires just 16 bytes, it needs to be 64B aligned. So we need to allocate 64B for each of them.

Manciukic commented on src/vmm/src/arch/aarch64/pvtime/mod.rs:
```diff
@@ -0,0 +1,117 @@
+        let base_addr = resource_allocator
+            .allocate_system_memory(
+                STEALTIME_STRUCT_MEM_SIZE * vcpu_count as u64,
+                16, // Q: We believe this to be 16, need confirmation.
```
> The base address must be 64 byte aligned and exist within a valid guest memory region.

https://www.kernel.org/doc/html/v5.8/virt/kvm/devices/vcpu.html#attribute-kvm-arm-vcpu-pvtime-ipa

Manciukic reviewed (commented):
I got confused by the KVM API as well! Passing a hva pointer to kvm is ok, but it needs to be 64B aligned.

The hint was in the KVM documentation:

> -EINVAL 	Base address not 64 byte aligned

https://www.kernel.org/doc/html/v5.8/virt/kvm/devices/vcpu.html#attribute-kvm-arm-vcpu-pvtime-ipa

DakshinD commented:
Thank you so much for your fast response! Totally missed the docs for 64 byte alignment, but that ended up fixing the bugs and I can successfully boot a microVM manually. If you think its a good time now, we can look to flesh out the implementation and "make it better".

Other than that, we believe the next step is adding unit/integration tests to make sure our device is working properly, as well as implementing persistence across snapshots. Any leads on how to do both of these things would be really helpful.

The lead we have been looking at for snapshotting is the `Persist` interface. 
- We store the `MicrovmState` in `save_snapshot`, so we need to first store our PVTime device somewhere in the vmm
- Then we implement `Persist` for PVTime, and store the `PVTimeState` we get from `save()` somewhere in the `MicrovmState` (could we extend MicrovmState to include this? or add this state to one of the other states already existing inside of it?)
- Finally, would need to write comprehensive tests to make sure this works

DakshinD commented:
Hi @Manciukic, I was just about to comment :), thanks for your response. We've started a bare bones implementation of snapshot persistence in the below commit. So if you also have the time to take a look at it, that would be awesome.

We are now trying to change the existing tests to include PVTime. Any input on our design choices we made would be appreciated again, specifically:
- Using either ONLY the base_addr or the hashmap in PVTime struct (we believe only the base_addr is necessary if the order of the vCPUs does not change across snapshots)
- Our restoration of the device inside `build_microvm_from_snapshot`
- Choosing to make the device an `Option<PVTime` inside the `struct Vmm`, and `Option<PVTimeState>` inside `struct MicrovmState`

Manciukic commented on src/vmm/src/builder.rs:
```diff
@@ -289,6 +296,17 @@ pub fn build_microvm_for_boot(
     attach_vmgenid_device(&mut vmm)?;
 
+    // Attempt to setup PVTime, continue if not supported
+    #[cfg(target_arch = "aarch64")]
+    if let Err(e) = setup_pv_time(&mut vmm, vcpus.as_mut()) {
+        match e {
+            StartMicrovmError::PVTimeNotSupported(e) => {
+                warn!("PVTime not supported: {}", e);
+            }
+            other => return Err(other),
+        }
+    }
+
```
With the changes proposed below, this would become
```
#[cfg(target_arch = "aarch64")]
vmm.pvtime = if PVTime::is_supported(&vcpus[0].kvm_vcpu.fd) {
        Some(setup_pv_time(&mut vmm, vcpus.as_mut())?)
    } else {
        warn!("PVTime is not supported by KVM. Steal time will not be reported to the guest.");
        None
};
```

Manciukic commented on src/vmm/src/arch/aarch64/pvtime/mod.rs:
```diff
@@ -0,0 +1,148 @@
+// Copyright 2025 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
nit: the path could be `src/vmm/src/arch/aarch64/pvtime.rs`

Manciukic reviewed (commented):
Overall looks good. I'd just move most of the pv_time specific logic into the new module to avoid cluttering the builder.rs even more.

We also need to bump the snapshot version and add a changelog entry, other than the integration tests.

Manciukic commented:
I tried to run the branch through our CI and there are a couple of issues: https://buildkite.com/firecracker/firecracker-pr/builds/13103#
 -  x86 doesn't compile, for example:
```
 error[E0433]: failed to resolve: could not find `aarch64` in `arch`
```
 - arm  has a few failures
```
integration_tests/build/test_clippy.py::test_rust_clippy[aarch64-unknown-linux-gnu] FAILED                                                               [ 12%]
integration_tests/build/test_clippy.py::test_rust_clippy[aarch64-unknown-linux-musl] FAILED                                                              [ 25%]
integration_tests/build/test_coverage.py::test_coverage FAILED                                                                                           [ 37%]
integration_tests/build/test_dependencies.py::test_unused_dependencies PASSED                                                                            [ 50%]
integration_tests/build/test_gdb.py::test_gdb_compiles PASSED                                                                                            [ 62%]
integration_tests/build/test_seccomp_no_redundant_rules.py::test_redundant_seccomp_rules PASSED                                                          [ 75%]
integration_tests/build/test_unittests.py::test_unittests FAILED                                                                                         [ 87%]
integration_tests/build/test_unittests.py::test_benchmarks_compile PASSED                                                                                [100%]
```
 - unit tests need to be updated with the new field
 - clippy is unhappy

But the good news is functional and performance tests passed :tada: 

To run these tests locally you can run:
```
./tools/devtool test -y -- integration_tests/build
```

DakshinD commented:
Managed to fix the errors reported in the build as well as clippy errors, would be awesome if it could be re-run.

One small question was if there was a way to view the code coverage tests. I see that a `.profraw` file gets generated, and I managed to find a way to convert it to `.profdata` and view it on a html file, but was wondering if there was an easier way to view which files/lines were covered.

We also have 2 random tests failing during integration tests, both having to do with `gicv2`
-`arch::aarch64::gic::gicv2::regs::dist_regs::tests::test_access_dist_regs stdout`
-`arch::aarch64::gic::gicv2::regs::tests::test_vm_save_restore_state stdout`

We started looking at unit/integration tests, as per your suggestions, and were hoping to get a small lead as to going about it. 

- For the unit tests in `src/vmm/src/builder.rs` as well as `test_microvm_state_snapshot`, we couldn't find a way to get the `Vec<Vcpus>` in context to setup the PVTime device. Is it necessary to mimic these existing tests for the pvtime device?
- I think we have a handle on the integration test with 2 cpus, so will go ahead writing that python test.

Manciukic commented:
> For the unit tests in `src/vmm/src/builder.rs` as well as `test_microvm_state_snapshot`, we couldn't find a way to get the `Vec<Vcpus>` in context to setup the PVTime device. Is it necessary to mimic these existing tests for the pvtime device?

It's difficult to unit test code that mostly performs syscalls (ioctls on vcpu fd in this case), so the policy to not over-complicate the codebase with mocks is to test these features through end-to-end integration tests only.
I think it would be enough to have unit tests inside `PVTime` to ensure the logic (excluding the `ioctl`s) is correct.
`builder.rs` is basically untesteable and is tested mostly

DakshinD commented:
Thanks for the clarification on that. We've added some very basic unit tests to PVTime and also 2 integration tests for basic checking and snapshotting. If there's any changes/next steps we need to take, please let us know. 

A re-run of the test suite would also be very helpful so we can nail down any bugs, as it was a bit buggy locally.

Manciukic commented:
> Thanks for the clarification on that. We've added some very basic unit tests to PVTime and also 2 integration tests for basic checking and snapshotting. If there's any changes/next steps we need to take, please let us know.

Can you also add a basic test that checks that the guest has enabled steal PV by looking at the kernel logs?
```
uvm.ssh.run("dmesg | grep 'stolen time PV'")
```
Make sure that only runs on aarch64.

> A re-run of the test suite would also be very helpful so we can nail down any bugs, as it was a bit buggy locally.

Done. x86 has the same issues as above: https://buildkite.com/firecracker/firecracker-pr/builds/13138#0196241a-77dd-4fe4-9023-e053f8004c2d

DakshinD commented:
Added that extra test and should've completely fixed the x86 errors, should be good for test suite re-run.

Manciukic commented on tests/integration_tests/functional/test_pvtime.py:
```diff
@@ -0,0 +1,146 @@
+    ]
+
+    # Ensure steal time persisted and continued increasing
+    persisted = sum(steal_before) + 100 <= sum(steal_after_snap)
+    increased = sum(steal_after_resume) > sum(steal_after_snap)
```
As above, this may be flaky (I'm even surprised it's working at the moment without any race!). For the purpose of our functional test we want to know that:
 - the value on restore is the same (approx) as the value at the time of snapshot
 - the steal time keeps increasing after the snapshot is restored

So the first condition needs to check that after_snap is between before and before+tolerance (we can remove the sleep above and keep the tolerance as 1s maybe). The second condition is ok.

Manciukic replied:
We're expecting steal time to increase by at least 1s, and we're checking exactly one second after. Any small measurement error may make it fail, but I don't think we need this kind of precision. We're not trying to test how accurate is the kernel, we just want to make sure that the steal time reporting is working after a snapshot resume.

Manciukic commented on tests/integration_tests/functional/test_pvtime.py:
as some of these tests also run on x86 (and I don't see any reason not to), maybe it's best to rename the test suite to `test_steal_time.py`

Manciukic commented on src/vmm/src/arch/aarch64/pvtime.rs:
```diff
@@ -0,0 +1,209 @@
+    pub fn register_all_vcpus(&self, vcpus: &mut [Vcpu]) -> Result<(), PVTimeError> {
+        // Register the vcpu with the pvtime device to map its steal time region
+        for (i, vcpu) in vcpus.iter().enumerate() {
+            #[allow(clippy::cast_possible_truncation)]
+            // We know vcpu_count is u8 according to VcpuConfig
+            self.register_vcpu(i as u8, &vcpu.kvm_vcpu.fd)?;
```
nit: I think a `try_into` would be nicer
```suggestion
            self.register_vcpu(i.try_into().expect("cpu index should be less than 256"), &vcpu.kvm_vcpu.fd)?;
```

DakshinD replied:
Avoiding this by using a u64 for vcpu_index

Manciukic commented on src/vmm/src/builder.rs:
```diff
@@ -12,13 +11,17 @@ use std::sync::{Arc, Mutex};
 use libc::EFD_NONBLOCK;
 use linux_loader::cmdline::Cmdline as LoaderKernelCmdline;
+#[cfg(target_arch = "aarch64")]
+use log::warn;
```
let's avoid conditionally importing non-arch-specific dependencies. We can probably just avoid the `use` and call `log::warn` directly.

Manciukic commented on src/vmm/src/arch/aarch64/pvtime.rs:
```diff
@@ -0,0 +1,209 @@
+#[derive(Debug)]
+pub struct PVTime {
+    /// Number of vCPUs
+    vcpu_count: u8,
```
I see down below that having this as `u8` adds some complication in handling the downcasts. As there's no tangible difference in using a `u64` I'd suggest to just use the bigger type and avoid all those downcast mess.

DakshinD replied:
I will go this route and use a u64 as the vcpu_count. I just assume that this places some sort of trust in the fact that vcpus.len() (during creation of pvtime) will never be above 256 and create a invalid vcpu index. Is any assert/check necessary here or is that situation not possible.

Manciukic replied:
I don't think it's necessary to assert the vcpu count as for what the pvtime it's concerned, as long as the vcpu fd is valid and the index is less than the total count, everything should work.

DakshinD commented on tests/integration_tests/functional/test_pvtime.py:
```diff
@@ -0,0 +1,146 @@
+
+    # Require increase in steal time
+    assert (
+        steal_after - steal_before >= 200
```
I also observed this was flaky, due to various other interrupts I believe. It was prevalent when we compared steal time on a per CPU basis as well. Due to the fact that "relaxing the constraint" or creating some tolerance could be finicky, I think checking a simple increase could be better.

DakshinD commented:
Thanks for the revisions @Manciukic. I've gone ahead and:
- changed `vcpu_index` to a u64 as per your suggestion and I think it cleans up the code nicely. It just assumes that `vcpus.len()` will not exceed 256 which I believe is a valid assumption?
- Fixed the integration tests to check for increase only, and also cleaned it up with helper function.

Please let me know if there is anything else on that end, and looking towards a merge, what would need to be done such as:
- changing snapshot version?
- changelog
- squashing my commits into ~3 substantial ones for clarity?

Manciukic commented:
> Thanks for the revisions @Manciukic. I've gone ahead and:
> 
>     * changed `vcpu_index` to a u64 as per your suggestion and I think it cleans up the code nicely. It just assumes that `vcpus.len()` will not exceed 256 which I believe is a valid assumption?

It's the assumption we already make in other parts of the code. I think it's fine.

> 
>     * Fixed the integration tests to check for increase only, and also cleaned it up with helper function.

Thanks!

> 
> 
> Please let me know if there is anything else on that end, and looking towards a merge, what would need to be done such as:
> 
>     * changing snapshot version?
> 
>     * changelog
> 
>     * squashing my commits into ~3 substantial ones for clarity?

Yup, all three are required for merging. I can't think of anything else at the moment, other than keeping the branch up-to-date with main.

I will take a deeper look next week after the Easter holidays.

DakshinD commented:
Just updated the snapshot version by minor version, added changelog, and attempted to rebase the commits but git freaked out so hopefully I did it right. Also apologies it seemed to auto-request reviews after I force pushed.

Manciukic commented on src/vmm/src/persist.rs:
```diff
@@ -152,7 +157,7 @@ pub enum CreateSnapshotError {
 
 /// Snapshot version
-pub const SNAPSHOT_VERSION: Version = Version::new(5, 0, 0);
+pub const SNAPSHOT_VERSION: Version = Version::new(6, 1, 0);
```
We only update the major version at every change.

```suggestion
pub const SNAPSHOT_VERSION: Version = Version::new(6, 0, 0);
```

Manciukic commented on tests/integration_tests/functional/test_steal_time.py:
```diff
@@ -0,0 +1,119 @@
+
+    # Ensure steal time persisted and continued increasing
+    tolerance = 1500  # 1.5 seconds tolerance for persistence check
+    persisted = steal_after_snap - steal_before < tolerance
```
We also want to assert it didn't go back.

```suggestion
    persisted = steal_before < steal_after_snap and steal_after_snap - steal_before < tolerance
```

Manciukic reviewed (commented):
mostly lgtm except a few minor things. Another small nit: please clean up the commit messages a little (make sure to mention what, why, and how).

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -548,6 +590,24 @@ pub fn setup_serial_device(
+/// Sets up the pvtime device.
+#[cfg(target_arch = "aarch64")]
+fn setup_pv_time(vmm: &mut Vmm, vcpus: &mut [Vcpu]) -> Result<PVTime, StartMicrovmError> {
+    use crate::arch::aarch64::pvtime::PVTime;
```
This import seems unnecessary

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -289,6 +295,17 @@ pub fn build_microvm_for_boot(
+            warn!("PVTime is not supported by KVM. Steal time will not be reported to the guest.");
+            None
+        };
+    }
```
This can be moved into `create_vmm_and_vcpus`

ShadowCurse commented on src/vmm/src/arch/aarch64/pvtime.rs:
```diff
@@ -18,7 +18,7 @@ pub const STEALTIME_STRUCT_MEM_SIZE: u64 = 64;
 pub struct PVTime {
     /// Number of vCPUs
-    vcpu_count: u8,
+    vcpu_count: u64,
```
Please move rust related fixes/changes into the first one, so they are not scattered. Let second commit be just a python integration test.

ShadowCurse commented on CHANGELOG.md:
```diff
@@ -10,6 +10,9 @@ and this project adheres to
 
 ### Added
 
+- [#5139](https://github.com/firecracker-microvm/firecracker/pull/5139): Added
```
Please move `CHANGELOG` update into a separate commit.

ShadowCurse commented on src/vmm/src/arch/aarch64/pvtime.rs:
```diff
@@ -0,0 +1,207 @@
+impl PVTime {
+    /// Helper function to get the IPA of the steal_time region for a given vCPU
+    fn get_steal_time_region_addr(&self, vcpu_index: u64) -> Result<GuestAddress, PVTimeError> {
+        if vcpu_index >= self.vcpu_count {
+            return Err(PVTimeError::InvalidVcpuIndex(vcpu_index));
+        }
```
This will also change return type to just `GuestAddress`

ShadowCurse reviewed (commented):
Overall looks reasonable. Couple minor fixes and it should be good to go.

roypat commented on src/vmm/src/arch/aarch64/pvtime.rs:
Since we have one pvtime memory area per vcpu, I think it makes more have a single field `pvtime_ipa: Option<GuestAddress>` inside each `struct KvmVcpu`, instead of one struct that manages PV time for all vcpus. Then a lot of the functions that are currently defined on struct PvTime but take vcpu[_index] as arguments will just become `&self` functions on struct KvmVcpu, and the whole setup would essentially become

```
let stealtime_mem = ResourceAllocator.allocate(...);

for vcpu in vcpus {
	if vcpu.supports_stealtime() { // does .has_device_attr()
        vcpu.enable_stealtime(stealtime_mem + i * STEALTIME_STRUCT_MEM_SIZE)?;  // does .set_device_attr()
    }
}
```

DakshinD replied:
Hi, thanks for looking over the PR. So from my understanding, you want to move methods such as `is_supported` and `register_vcpu` inside `KvmVcpu`? 

This does make sense to me, but then I believe some code might become redundant, as if we add `pvtime_ipa` to `KvmVcpu` and persist that, we would also need to persist the `base_ipa` inside `PVTime` to restore the system memory. Unless we do one or the other, maybe get rid of `PVTime` entirely and handle the restore logic somewhere else?

roypat replied:
I think we can persist the `pvtime_ipa` inside each `struct VcpuState` as you say. I don't think there is a need to persist the `base_ipa` at all in this case, so `PVTime` would be gone entirely

DakshinD replied:
Thanks for the clarification, as this is a large change to my existing PR that I've been working on, wanted to confirm the "go-ahead" on this by listing out my thought process.

- Move methods such as support checking and registering ipa into `KvmVcpu`
- `setup_pv_time` will change to allocating system memory itself and iterating over all vcpus and registering them as you proposed in your code snippet above
- The logic to restore the pvtime device in `build_microvm_from_snapshot` disappears as we will move the `ipa` per vcpu into the `VcpuState` that gets saved in `Vcpu::paused` after the vcpu is sent an event to save in `Vmm::save_vcpu_states`. This requires me to change the `save_state` and `restore_state` (re-register ipa in here) methods of `KvmVcpu`

If these changes are ok, I will go ahead and get them done ASAP. One question I did have was what would happen to the `PVTime` unit tests. Would there be no unit tests, just integration tests now? Or would I move them into `KvmVcpu`?

roypat replied:
Yes, that sounds good to me, thanks! As for the unit tests, yeah, it does seems like the code they're currently testing would just go away, so just having integration tests seems fine to me.

DakshinD commented:
Thanks for the PR review earlier. 

I have gone ahead and:
- Addressed @roypat design changes. I have gone ahead and got rid of `PVTime` device, moved the logic into `KvmVcpu`, and just left some setup logic inside builder.rs. 
- I have cleaned up the commit messages, as well as fixed the small nit on the integration test as @Manciukic said.
- I have separated the commits into rust and integration test commits, as well as a separate one for the Changelog as @ShadowCurse  said.

I've double checked that style, clippy, steal time tests pass, etc. Any feedback on this new design would be welcome, as well as a run of the test-suite if possible. I believe (hopefully) that it should build fine on x86 as well. 

Thank you!

roypat commented on src/vmm/src/builder.rs:
```diff
@@ -552,6 +573,44 @@ pub fn setup_serial_device(
+fn allocate_pvtime_region(
+    vmm: &mut Vmm,
+    vcpu_count: usize,
+    policy: vm_allocator::AllocPolicy,
```
If you drop the call to this on the restore path, you can also drop this `policy` argument (and just inline this into `setup_pvtime` probably).

roypat reviewed (commented):
Awesome, thanks for doing that change! Everything looks a lot simpler now :D 

Just a small comment about the restore code, but then this looks good to go from me too!

Manciukic commented on src/vmm/src/builder.rs:
```diff
@@ -449,6 +460,16 @@ pub fn build_microvm_from_snapshot(
         }
     }
 
+    // Need to alloc sys mem for steal time regions before vcpu setup
+    #[cfg(target_arch = "aarch64")]
+    if let Some(pvtime_ipa) = vcpus[0].kvm_vcpu.pvtime_ipa {
+        allocate_pvtime_region(
+            &mut vmm,
+            vcpus.len(),
+            vm_allocator::AllocPolicy::ExactMatch(pvtime_ipa.0),
+        )?;
+    }
```
Yeah, that seems to be the pattern followed for vmgenid and mmio devices. Checking again, this allocator is used 3 times (`mptable`, `acpi`, `vmgenid`), but only `vmgenid` restores its state, while the other 2 places don't, which means it's already out-of-sync :(
Restoring the allocator state seems like a better option than remembering to go back and allocate again all places (but it's also out of scope). I'll open an issue for that.

roypat replied:
hahahaha, if we already aren't doing this consistently, I have a weak opinion in favor of not doing it here either xP

Manciukic commented:
One more issue on x86 :( 

```
error[E0432]: unresolved import `crate::arch::VcpuArchError`
  --> src/vmm/src/builder.rs:23:39
   |
23 | use crate::arch::{ConfigurationError, VcpuArchError, configure_system_for_boot, load_kernel};
   |                                       ^^^^^^^^^^^^^
   |                                       |
   |                                       no `VcpuArchError` in `arch`
   |                                       help: a similar name exists in the module: `KvmArchError`
warning: unused import: `vm_memory::GuestAddress`
  --> src/vmm/src/builder.rs:17:5
   |
17 | use vm_memory::GuestAddress;
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default
```

DakshinD commented:
@Manciukic Just pushed that fix for x86, sorry about that. Gonna try getting the cross-compilation working locally so I can test that soon.

DakshinD commented:
Thanks for pointing that out. It was telling me `the `x86_64-unknown-linux-gnu` target may not be installed`, so I would do `rustup target add x86_64-unknown-linux-gnu` and it would tell me everything was fine. But I had to run it with sudo for it to actually download the needed things :( Oversight on my part
